coproc: Move materialized partitions

graphcareful opened the pull request:
This pull request adds the ability for redpanda to move materialized partitions. This will not be able to be manually controlled, however will occur whenever a materialized parent topic is moved. The materialized partition will move to the exact same node/shard as its parent topic.

Changes in [force-push](https://github.com/vectorizedio/redpanda/compare/013820374467ba8e2fef5addf9a0c917fecf2fcd..a6f19951810c01be8cd0a5bc4bdd092ad77b8538)
- Remove stale code
- Removed test harness that has already been checked into dev
- Fixed bug with partitions that are restarting across nodes, restart was not occurring

Changes in [force-push](https://github.com/vectorizedio/redpanda/compare/a6f19951810c01be8cd0a5bc4bdd092ad77b8538..663025a9d9fa8e2046e0a66d3fae4908f1d2980a)
- Rebased against dev to obtain the recently checked in `coproc_cluster_fixture`

Chagnes in [force-push](https://github.com/vectorizedio/redpanda/compare/663025a9d9fa8e2046e0a66d3fae4908f1d2980a..7b6338929cb5cc11eff2817b1c376e7aedb73052)
- Fixed bug in new wasm partition movement unit test

Changes in [force-push](https://github.com/vectorizedio/redpanda/compare/7b6338929cb5cc11eff2817b1c376e7aedb73052..423385fc5c21e683fc1d5d994a4ee56733a42b91)
- Removed x-shard materialized partition move optimization
- Fixed bug where processing post move didn't start at offset 0
- Fixed bug where topic config updates during move didn't propagate to materialized topics, resulting in stale metadata.
- Fixed bug where admin-api wasn't reporting materialized topics.
- New ducktape test

Changes in [force-push](https://github.com/vectorizedio/redpanda/compare/423385fc5c21e683fc1d5d994a4ee56733a42b91..423385fc5c21e683fc1d5d994a4ee56733a42b91)
- Reverting erraneous force-push

Changes in [force-push](https://github.com/vectorizedio/redpanda/compare/423385fc5c21e683fc1d5d994a4ee56733a42b91..ebf9f17a972dac0fc2e065b65a9f5cc80907170c)
- Implemented Ben's C++ suggestions

Changes in [force-push](https://github.com/vectorizedio/redpanda/compare/ebf9f17a972dac0fc2e065b65a9f5cc80907170c..8c72568505565b2fff8d0966595bf76170b6eaf1)
- Rebased dev to resolve one line conflict in `coproc/api.h`

Changes in [force-push](https://github.com/vectorizedio/redpanda/compare/8c72568505565b2fff8d0966595bf76170b6eaf1..c8722ce73a7ebfe78bde919edaa3ef95f4eb1101)
- Fixed all of the bugs mentioned in my personal review comments below

Changes in [force-push](https://github.com/vectorizedio/redpanda/compare/c8722ce73a7ebfe78bde919edaa3ef95f4eb1101..aa2e6b28df0126829fa93fc31ade899fbb2e5600)
- Replaced use of `ss::circular_buffer<>` within reconciliation backend

Changes in [force-push](https://github.com/vectorizedio/redpanda/compare/aa2e6b28df0126829fa93fc31ade899fbb2e5600..b2754dae38ed993a367887b181e3e57b7db9d2ae)
- Rebased against dev to resolve conflicts with modifications to ducktape tests 

Changes in [force-push](https://github.com/vectorizedio/redpanda/compare/b2754dae38ed993a367887b181e3e57b7db9d2ae..4b0529758d1d2f35c12dd2d8615f0960bf119dff)
- Modified git commit messages
- Replaced assert with log
- Fixed issue where code was edited in commit that had stated only a move was performed
- Added a new field in 'partition_summary' that denotes if a partition is materialized or not
- Added a new test that crashes a node in the middle of a materialized partition move

Changes in [force-push](https://github.com/vectorizedio/redpanda/compare/4b0529758d1d2f35c12dd2d8615f0960bf119dff..82eb39fbcaa5be4a9b366e28d4ba241a34bf316b)
- Removed @ignore directive on test that crashed a node during move now that #3558 has been merged

graphcareful commented:
PR is missing one final commit, a new ducktape test. Plan is to push the PR now to get feedback earlier while test is asynchronously finished.

VadimPlh commented on src/v/cluster/topic_table.cc:
```diff
@@ -161,10 +161,13 @@ topic_table::apply(create_partition_cmd cmd, model::offset offset) {
 ss::future<std::error_code>
 topic_table::apply(move_partition_replicas_cmd cmd, model::offset o) {
     auto tp = _topics.find(model::topic_namespace_view(cmd.key));
-    if (tp == _topics.end() || !tp->second.is_topic_replicable()) {
```
Should we use another error instead `topic not exists` for not replicable topic?

VadimPlh commented on src/v/coproc/script_context.cc:
```diff
@@ -153,12 +186,36 @@ script_context::process_send_write(rpc::transport* t) {
+          "Wanting to swap a different ctx for same input ntp");
+    }
+    auto [itr, _] = _updates.emplace(
+      ntp, std::make_unique<insert_update>(std::move(rctx)));
```
You got `&&` for rctx, but use move inside
Maybe change `read_context&& rctx` to `read_context rctx`

graphcareful replied:
Thats fine though right? You can move rvalues

VadimPlh commented on src/v/coproc/api.cc:
```diff
@@ -58,10 +59,11 @@ ss::future<> api::start() {
     co_await _pacemaker.invoke_on_all(&coproc::pacemaker::start);
-    _listener = std::make_unique<wasm::event_listener>();
 
+    _listener = std::make_unique<wasm::event_listener>(_as);
```
I think we should not pass as to event_listener from coproc::api.
event_listener is not only coproc listener. In future it will listen data_policy event, inline_wasm events...

graphcareful replied:
I think we need to modify how the event listener is instantiated. Something like
```
if(enable_coproc() || enable_v8()) {
  _listener = std::make_unique<wasm::event_listener(_as);
}
```

VadimPlh commented on src/v/coproc/script_database.h:
```diff
@@ -0,0 +1,112 @@
+    const const_iterator cend() const { return _db.cend(); }
+
+    /// ... by topic/namespace
+    bool exists(const model::topic_namespace& tn) const {
```
`model::topic_namespace_view `?

mmaslankaprv commented on src/v/coproc/script_dispatcher.h:
```diff
@@ -72,6 +75,9 @@ class script_dispatcher {
     /// across all shards
     ss::sharded<pacemaker>& _pacemaker;
 
+    /// TODO: leave impotant comment here
```
this comment needs an update

mmaslankaprv commented:
I think we can add some materialized topics to `node_operations_fuzzy` test this way we will how the feature is working during node operations.

dotnwat reviewed (commented):
I haven't done a full review yet but a few things I see

1. the cover letter can really benefit from a description of how this works. how does it integrate with the existing partition movement? what are the important edge cases? etc...
2. related to (1) I didn't see any invocations of `partition_manager::register_manage_notification` which I would have expected for subscribing to events of a parent/source partition.
3. I agree with @mmaslankaprv that this should integrate in some way with existing partition movement tests. in addition to his suggestion about integration with `node_operations_fuzzy_test` this should also integrate with the `partition_movement_test` unit test and the `partition_movement` duck tape test. in the later two cases I think you could imagine arranging for materialized partitions to be associated with the partitions being tested and then make assertions about the fate of the materialized partitions. those two tests (unit and duck tape) took a long time to stabilize in the system, so they are an excellent thing to piggy back on for stability checks.

BenPope commented on src/v/coproc/script_context.h:
```diff
@@ -25,7 +26,7 @@
 namespace coproc {
 
-/// Raised when futures enqueued by \ref remove_output were failed to be
+/// Raised when futures enqueued by \ref an async update were failed to be
```
nit: The `\ref` is invalid

BenPope commented on src/v/coproc/pacemaker.cc:
```diff
@@ -173,6 +173,8 @@ std::vector<errc> pacemaker::add_source(
                 "expected: {}",
                 e.get_id(),
                 id);
+              /// TODO: Poses interesting issue, now that there is a script
```
Can you add your name to this TODO

BenPope commented on src/v/coproc/api.cc:
```diff
@@ -58,10 +59,11 @@ ss::future<> api::start() {
+    _dispatcher = std::make_unique<wasm::script_dispatcher>(_pacemaker, _as);
     _wasm_async_handler = std::make_unique<coproc::wasm::async_event_handler>(
-      _listener->get_abort_source(), std::ref(_pacemaker));
+      std::ref(*_dispatcher));
```
I know `start` should only be called once, but I think it's worth asserting that a bunch of these pointers are null at the beginning of this function.

BenPope commented on src/v/coproc/script_database.h:
```diff
@@ -0,0 +1,112 @@
+#include <vector>
+
+namespace coproc::wasm {
+/// In memory cache of all coprocessors of all coprocessors
```
nit: Typo?

BenPope commented on src/v/coproc/script_database.h:
```diff
@@ -0,0 +1,112 @@
+
+    /// Querying
+    /// ... by scripts
+    bool exists(script_id id) { return _db.find(id) != _db.end(); }
```
```suggestion
    bool exists(script_id id) { return _db.contains(id); }
```

BenPope commented on src/v/coproc/script_database.h:
```diff
@@ -0,0 +1,112 @@
+    /// Querying
+    /// ... by scripts
+    bool exists(script_id id) { return _db.find(id) != _db.end(); }
+    const const_iterator find(script_id id) const { return _db.find(id); }
+    const const_iterator cbegin() const { return _db.cbegin(); }
+    const const_iterator cend() const { return _db.cend(); }
```
Why return a `const const_iterator` by value?

There are other examples of this below.

BenPope commented on src/v/coproc/script_database.h:
```diff
@@ -0,0 +1,112 @@
+
+    /// ... by topic/namespace
+    bool exists(model::topic_namespace_view tnv) const {
+        return _by_input.find(tnv) != _by_input.cend();
```
```suggestion
        return _by_input.contains(tnv);
```

BenPope commented on src/v/coproc/script_database.h:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    /// Addition
+    bool add_script(
+      script_id id, iobuf source, std::vector<topic_namespace_policy> inputs) {
+        script_metadata meta{.source = std::move(source), .inputs = inputs};
+        auto [_, success] = _db.emplace(id, std::move(meta));
+        if (!success) {
+            return false;
+        }
+        for (auto& input : inputs) {
+            auto found = _by_input.find(input.tn);
+            if (found == _by_input.end()) {
+                absl::node_hash_set<script_id> ns;
+                ns.emplace(id);
+                _by_input.emplace(std::move(input.tn), std::move(ns));
+            } else {
+                found->second.emplace(id);
+            }
+        }
+        return true;
+    }
```
It's tricky to make this exception safe. Can you explain the expected usage patterns of this `script_database`?

In the past I've used `boost::multi_index_container`.

graphcareful replied:
Nice food for thought, I will think about other options. To answer the question this class has 2 maps:
1. map<id -> list<topic>>
2. map<topic, list<id>>

The purpose of this is to lookup what input topics a script is interested, or do the inverse, depending on an input topic lookup which scripts were registered for this. The second map is particularly useful for partition movement, because when a source topic is moved/re-inserted it is necessary to know which co-processors care about this event.

graphcareful replied:
Maybe @VadimPlh can help here. I'm noticing I actually never invoke `script_database::find(script_id)`. I think I included it as I thought he might want to perform lookups by `script_id`. @VadimPlh do you need to do that for data policies ever? I remember seeing a map of scripts in one of your recent PRs

BenPope replied:
If it's a lookup that runs rarely, and the number of entries is fewer than say, 100K a linear search is often good enough, and that might inform whether you need stable iterators on the primary data structure.

BenPope commented on src/v/coproc/partition_manager.cc:
```diff
@@ -84,6 +84,20 @@ ss::future<> partition_manager::remove(const model::ntp& ntp) {
+    auto partition = get(ntp);
+
+    if (!partition) {
+        return ss::make_exception_future<>(std::invalid_argument(fmt::format(
```
Should there be an exception derived from `coproc::exception` for this?

BenPope commented on src/v/coproc/partition_manager.cc:
```diff
@@ -84,6 +84,20 @@ ss::future<> partition_manager::remove(const model::ntp& ntp) {
 }
 
+ss::future<> partition_manager::shutdown(const model::ntp& ntp) {
+    auto holder = _gate.hold();
```
This gate isn't held during `do_shutdown`

BenPope commented on src/v/coproc/reconciliation_backend.cc:
```diff
@@ -18,25 +18,72 @@
+namespace {
+
+std::optional<ss::shard_id>
+new_shard(model::node_id self, std::vector<model::broker_shard> new_replicas) {
```
Copying this vector seems unnecessary.

dotnwat commented:
seeing a few test issues

```
test_id:    rptest.tests.wasm_partition_movement_test.WasmPartitionMovementTest.test_dynamic
--
  | status:     FAIL
  | run time:   1 minute 18.582 seconds
  |  
  |  
  | TimeoutError('')
  | Traceback (most recent call last):
  | File "/usr/local/lib/python3.8/dist-packages/ducktape/tests/runner_client.py", line 135, in run
  | data = self.run_test()
  | File "/usr/local/lib/python3.8/dist-packages/ducktape/tests/runner_client.py", line 215, in run_test
  | return self.test_context.function(self.test)
  | File "/root/tests/rptest/tests/wasm_partition_movement_test.py", line 124, in test_dynamic
  | wait_until(topic_created, timeout_sec=30, backoff_sec=2)
  | File "/usr/local/lib/python3.8/dist-packages/ducktape/utils/util.py", line 58, in wait_until
  | raise TimeoutError(err_msg() if callable(err_msg) else err_msg) from last_exception
  | ducktape.errors.TimeoutError
```

graphcareful commented:
> seeing a few test issues
> 
> ```
> test_id:    rptest.tests.wasm_partition_movement_test.WasmPartitionMovementTest.test_dynamic
> --
>   | status:     FAIL
>   | run time:   1 minute 18.582 seconds
>   |  
>   |  
>   | TimeoutError('')
>   | Traceback (most recent call last):
>   | File "/usr/local/lib/python3.8/dist-packages/ducktape/tests/runner_client.py", line 135, in run
>   | data = self.run_test()
>   | File "/usr/local/lib/python3.8/dist-packages/ducktape/tests/runner_client.py", line 215, in run_test
>   | return self.test_context.function(self.test)
>   | File "/root/tests/rptest/tests/wasm_partition_movement_test.py", line 124, in test_dynamic
>   | wait_until(topic_created, timeout_sec=30, backoff_sec=2)
>   | File "/usr/local/lib/python3.8/dist-packages/ducktape/utils/util.py", line 58, in wait_until
>   | raise TimeoutError(err_msg() if callable(err_msg) else err_msg) from last_exception
>   | ducktape.errors.TimeoutError
> ```

Yes this PR needs #3264 to work correctly with multi-node tests

dotnwat commented:
what was force pushed?

dotnwat commented:
@graphcareful 

```
22 - coproc_fixture_rpunit (Failed)
```

graphcareful commented:
Bens suggestions were force pushed, investigating the CI failure now

graphcareful commented on src/v/coproc/reconciliation_backend.cc:
```diff
@@ -151,7 +201,99 @@ reconciliation_backend::process_update(update_t delta) {
+          "due to this node not having yet consumed all coprocessors from "
+          "internal coprocessor topic.",
+          ntp);
+        /// TODO: However it should start from offset 0
```
This has been addressed, update comments / and or commit messages

graphcareful commented on tests/rptest/tests/wasm_partition_movement_test.py:
```diff
@@ -0,0 +1,140 @@
+            script.get_artifact(self._build_tool.work_dir), script.name,
+            "ducktape")
+
+    def _verify_materialized_assignments(self, topic, partition, assignments):
```
Maybe _do_move_verify can be re-used here to reduce code duplication

graphcareful commented on tests/rptest/tests/wasm_partition_movement_test.py:
```diff
@@ -0,0 +1,140 @@
+        wait_until(topic_created, timeout_sec=30, backoff_sec=2)
+
+        t, p = self._grab_input(spec.name)
+        for _ in range(2):
```
We can increase this number, maybe to 5 runs

dotnwat commented:
I was able to run the ducktape wasm partition movement test the first time I tried. But for each subsequent run I've tried I'm getting a timeout creating the the source topic:

```
INFO:2021-12-22 01:57:22,634]: RunnerClient: rptest.tests.wasm_partition_movement_test.WasmPartitionMovementTest.test_dynamic: Data: None
================================================================================
SESSION REPORT (ALL TESTS)
ducktape version: 0.8.8
session_id:       2021-12-22--004
run time:         1 minute 9.350 seconds
tests run:        1
passed:           0
failed:           1
ignored:          0
================================================================================
test_id:    rptest.tests.wasm_partition_movement_test.WasmPartitionMovementTest.test_dynamic
status:     FAIL
run time:   1 minute 9.323 seconds
    TimeoutError('')
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/ducktape/tests/runner_client.py", line 135, in run
    data = self.run_test()
  File "/usr/local/lib/python3.8/dist-packages/ducktape/tests/runner_client.py", line 215, in run_test
    return self.test_context.function(self.test)
  File "/root/tests/rptest/tests/wasm_partition_movement_test.py", line 124, in test_dynamic
    wait_until(topic_created, timeout_sec=30, backoff_sec=2)
  File "/usr/local/lib/python3.8/dist-packages/ducktape/utils/util.py", line 58, in wait_until
    raise TimeoutError(err_msg() if callable(err_msg) else err_msg) from last_exception
ducktape.errors.TimeoutError
```

30 seconds is a long time, so it probably doesn't need to be increased.

One thing I see in the broker logs is

```
INFO  2021-12-22 01:57:17,498 [shard 2] coproc - script_context_backend.cc:360 -                                                                                                                                                                                                      
626 Error while processing reply: coproc::follower_create_topic_exception (Follower                                                                                                                                                                                                       
627 of source topic {ns: {kafka}, topic: {topic}} attempted to created materialzied                                                                                                                                                                                                       
628 topic {ns: {kafka}, topic: {topic._identity_output_}} before leader partition                                                                                                                                                                                                         
629 had a chance to, sleeping 1s...)
```

Which seems related. It's odd that it worked the very first time, but not after that. I wonder if there is some state being left around that is causing problems?

graphcareful commented on src/v/coproc/reconciliation_backend.cc:
```diff
@@ -88,7 +130,9 @@ reconciliation_backend::process_events_for_ntp(
-        if (err == errc::partition_not_exists) {
+        if (
+          d.type == update_t::op_type::add_non_replicable
+          && err == errc::partition_not_exists) {
```
Confusingly if we observe the following op types:
```
add -> add materialized -> update
```
Then on restart, since its impossible to synchronize around the update and add events (because they are handled by `controller_backend`) when processing `add_materialized` occurs there's a possibility the source topic doesn't exist because the `controller_backend` already moved it across shards or onto another node.

There is an order dependency between the processing of the controller log between this class and `controller_backend`. Moreso they can be processing the controller log at totally different offset. Maybe it would be cleaner to put all events event add / delete coprocessor ones in the controller log, but aside from that a solution to the aforementioned problem would be to just verify if this event can be ignored by checking if the partition assignments for the source topic and materialized topics (found in topics table) and equal. If they aren't the processing of the op_t can be ignored, safely assuming there is a future `update` request incoming.

For example:
```
add (always ignored by reconciliation_backend) -> add materialized (sees non matching partition assignments, skips) -> update (attempts to create materialized partition)
```
and things continue to work as normal, since an update in any case always creates a materialized partition.

graphcareful replied:
To add onto this, when processing for any op_t returns `partition_not_exists` forward progress cannot be made until the op_t is processed. Possibilities for infinite looping could occur here

graphcareful commented on src/v/coproc/reconciliation_backend.cc:
```diff
@@ -88,7 +130,9 @@ reconciliation_backend::process_events_for_ntp(
         auto err = co_await process_update(d);
         vlog(coproclog.info, "partition operation {} result {}", d, err);
-        if (err == errc::partition_not_exists) {
+        if (
```
There's a bug with this method. In the event there's a retry it continues processing the next event anyway. Out of order processing will occur in this scenario

graphcareful commented on src/v/coproc/reconciliation_backend.cc:
```diff
@@ -151,7 +201,99 @@ reconciliation_backend::process_update(update_t delta) {
+      wasm::script_database_main_shard, [ntp](wasm::script_database& sdb) {
+          return grab_metadata(sdb, model::topic_namespace_view{ntp});
+      });
+    if (meta.empty()) {
```
I think the best solution would be to process all coprocessor add/remove events before starting the reconciliation backends loop. This means if the condition above (`if(meta.empty()`) has been reached, it is truly an event that can be ignored, it is guaranteed that the script was eventually deleted, instead of not knowing if the script has or has not been deleted. Which would mean to log the event and "wait" for the coprocessor to be processed off of the coprocessor internal topic. Which is what this source at the time of writing this comment is doing.

graphcareful commented on src/v/coproc/reconciliation_backend.cc:
```diff
@@ -151,7 +201,99 @@ reconciliation_backend::process_update(update_t delta) {
+
+    /// Finally remove log from disk
+    for (const auto& ntp : ntps) {
+        co_await _coproc_pm.local().remove(ntp);
```
Hold onto the partition before removal

graphcareful commented on src/v/coproc/reconciliation_backend.h:
```diff
@@ -44,11 +44,10 @@ class reconciliation_backend {
-    ss::future<std::vector<update_t>>
-      process_events_for_ntp(model::ntp, std::vector<update_t>);
+    using events_cache_t
+      = absl::node_hash_map<model::ntp, ss::circular_buffer<update_t>>;
```
@dotnwat opinions on using a circular buffer here?

graphcareful replied:
Resolving as recent force push reverts this

graphcareful commented on src/v/redpanda/admin_server.cc:
```diff
@@ -1072,27 +1074,35 @@ void admin_server::register_partition_routes() {
-                  return acc;
-              })
-            .then([](std::vector<summary> partitions) {
+          auto get_summaries = [](auto& partition_manager) {
```
For reviewers, here is a place where "coproc" code is invoked even when the flag is off. Shouldn't be an issue since the `coproc::partition_manager` should always be empty in those cases anyway.

mmaslankaprv commented on src/v/cluster/topic_table.cc:
```diff
@@ -193,6 +201,32 @@ topic_table::apply(move_partition_replicas_cmd cmd, model::offset o) {
+        for (const auto& cs : found->second) {
+            auto [_, success] = _update_in_progress.insert(
+              model::ntp(cs.ns, cs.tp, current_assignment_it->id));
+            vassert(
```
This assert message would benefit from more context.

dotnwat commented:
@graphcareful your branch looks like it using some outdated ducktape apis which is leading to some of the CI failures.

mmaslankaprv commented on src/v/coproc/reconciliation_backend.cc:
```diff
@@ -36,94 +62,112 @@ reconciliation_backend::reconciliation_backend(
+    if (deltas.empty()) {
+        /// Prevent spin looping when theres nothing to process
+        try {
+            co_await ss::sleep_abortable(1s, _as);
```
why we decided to remove the timer and use sleeps instead ?

graphcareful replied:
There was a bug in the previous implementation where out of order processing could occur. To make things simpler i decided to have one fiber that just ingests until theres no data, sleeping 1s if there is, then another context fills a vector with `updates` when they arrive from the topics table. No mutexs and no timers are now needed.

mmaslankaprv commented on src/v/coproc/reconciliation_backend.cc:
```diff
@@ -22,6 +22,32 @@
+namespace {
+
+using delta_op_type = cluster::topic_table_delta::op_type;
+bool is_replicable_event(delta_op_type t) {
```
should this be named `is_non_replicable_event` ?

mmaslankaprv commented on src/v/coproc/reconciliation_backend.cc:
```diff
@@ -36,94 +62,112 @@ reconciliation_backend::reconciliation_backend(
+        } else {
+            vlog(
+              coproclog.warn,
+              "There were recoverable errors when processing events, "
```
there is only one error causing this update to be retried, maybe we should add this information to the log entry ?

mmaslankaprv commented on src/v/coproc/partition_manager.cc:
```diff
@@ -77,6 +77,7 @@ ss::future<> partition_manager::remove(const model::ntp& ntp) {
           ntp));
     }
 
+    vlog(coproclog.info, "Removing materialized log: {}", ntp);
```
this seems like unrelated changes.

graphcareful replied:
Yeah i wasn't sure where to put it, but it makes debugging much easier, should i make a separate commit?

mmaslankaprv commented on src/v/coproc/reconciliation_backend.cc:
```diff
@@ -195,7 +223,72 @@ reconciliation_backend::process_update(model::ntp, update_t delta) {
+    co_return make_error_code(errc::success);
+}
+
+ss::future<std::error_code> reconciliation_backend::process_restart(
```
Would it be possible to add an overview of partition movement logic in the commit message ?

mmaslankaprv commented on src/v/coproc/reconciliation_backend.cc:
```diff
@@ -225,10 +353,14 @@ reconciliation_backend::create_non_replicable_partition(
+        auto src_partition = _cluster_pm.local().get(src_ntp);
         if (!src_partition) {
-            co_return errc::partition_not_exists;
+            co_return stale_create_non_replicable_partition_request(
```
can `stale_create_non_replicable_partition_request` return an error code ?

graphcareful replied:
No its either a stale request or not, so boolean

graphcareful commented:
> I would love to add a test where nodes are being decommissioned and added with non replicable topics in the cluster. I am a bit worried about how will it operate in the face of multiple partitions movements and failures

I did write a test that injected failures during moves, but it failed half of the time due to the random partition assignment choosing a x-core move and thus hitting that x-core move bug we observed in dev.

dotnwat reviewed (commented):
this is seeming pretty good to me. just some questions and some places where some extra comments would be useful.

can you also comment in the PR about what parts of coproc / co-proc adjacent things are reachable even when coproc feature is disabled?

mmaslankaprv commented:
> > I would love to add a test where nodes are being decommissioned and added with non replicable topics in the cluster. I am a bit worried about how will it operate in the face of multiple partitions movements and failures
> 
> I did write a test that injected failures during moves, but it failed half of the time due to the random partition assignment choosing a x-core move and thus hitting that x-core move bug we observed in dev.

@graphcareful i think i fixed the problem you describe here: #3558, do you think we can try running this test with that fix included ?

graphcareful commented on src/v/coproc/reconciliation_backend.cc:
```diff
@@ -36,94 +62,112 @@ reconciliation_backend::reconciliation_backend(
+    if (deltas.empty()) {
+        /// Prevent spin looping when theres nothing to process
+        try {
+            co_await ss::sleep_abortable(1s, _as);
```
So that exists (at the begging of `start()`), but since that method isn't futurized a mechanism needs to be in-place to ensure in-order processing of these events.

Previously I was using the trigger as a mechanism to start processing, and if processing is already started, just start a timer to re-start sometime in the future. But I think it might be harder to get that right, as i found some bugs with out of order processing.

With this new implementation, when the delta event arrives, it just places the update in a vector of to-be-processed events, where this main reconciliation fiber will eventually consume it. No timers, no mutexes and a simpler thread of context to understand, i.e. the loop cannot continue until the current update completes without a retryable error.

graphcareful commented on src/v/coproc/reconciliation_backend.cc:
```diff
@@ -167,24 +189,30 @@ reconciliation_backend::fetch_and_reconcile(events_cache_t deltas) {
+        return create_non_replicable_partition(
+          delta.ntp, rev, delta.new_assignment.replicas);
     case op_t::del_non_replicable:
-        return delete_non_replicable_partition(delta.ntp, rev).then([] {
-            return std::error_code(errc::success);
-        });
-    case op_t::add:
-    case op_t::del:
+        return delete_non_replicable_partition(delta.ntp, rev);
     case op_t::update:
+        return process_shutdown(
+          delta.ntp, ntp, rev, std::move(delta.new_assignment.replicas));
     case op_t::update_finished:
+        return process_restart(
+          delta.ntp, ntp, rev, delta.new_assignment.replicas);
+    case op_t::add:
+    case op_t::del:
```
Yeah totally agree +1

graphcareful commented on src/v/cluster/topic_table.cc:
```diff
@@ -240,6 +278,18 @@ topic_table::apply(finish_moving_partition_replicas_cmd cmd, model::offset o) {
+              _update_in_progress.erase(
+                model::ntp(cs.ns, cs.tp, cmd.key.tp.partition))
+                > 0,
+              "state consistency issue non_replicable moved topic");
```
Will add more context and remove assertion replacing with error log

graphcareful commented on src/v/cluster/topic_table.cc:
```diff
@@ -240,6 +278,18 @@ topic_table::apply(finish_moving_partition_replicas_cmd cmd, model::offset o) {
+                > 0,
+              "state consistency issue non_replicable moved topic");
+        }
+    }
```
that it was expected to be in the `update_in_progress` collection due to a partition move event, but was then never found there.

graphcareful commented on src/v/cluster/topic_table.cc:
```diff
@@ -193,6 +201,32 @@ topic_table::apply(move_partition_replicas_cmd cmd, model::offset o) {
+            auto sfound = _topics.find(cs);
+            vassert(
+              sfound != _topics.end(),
+              "Non replicable topic must exist: {}",
```
Yup, the hierarchy map is a mapping of parent->children topics. Will add comments, but to explain here:
1. lookup is performed
2. iterate on all children
3. for each child update the assignments to match the parents

graphcareful commented on src/v/redpanda/admin_server.cc:
```diff
@@ -1074,27 +1076,35 @@ void admin_server::register_partition_routes() {
+                        p.core = ss::this_shard_id();
+                        partitions.push_back(std::move(p));
+                    }
+                    return partitions;
```
Good idea, i like it

graphcareful commented:
> does it make sense to return an error if coproc is not enabled?

The `coproc::partition_manager` is the only coproc entity that will exist when the flag is enabled. In this case it should always be empty and always return no items.

graphcareful commented on src/v/coproc/script_context.cc:
```diff
@@ -128,6 +110,50 @@ ss::future<> script_context::remove_output(
 }
 
+ss::future<ss::stop_iteration>
+script_context::process_send_write(rpc::transport* t) {
```
Looks like its kept alive as long as the `reconnect_transport` instance is alive. And one of those is kept alive by the pacemaker, per shard. https://github.com/vectorizedio/redpanda/blob/dev/src/v/coproc/shared_script_resources.h#L44

graphcareful commented on tests/rptest/tests/wasm_partition_movement_test.py:
```diff
@@ -0,0 +1,199 @@
+        # Since the identity copro was deployed, contents of logs should be identical
+        assert set(self.records_consumed) == set(self.result_data)
+
+    @ignore  # https://github.com/vectorizedio/redpanda/pull/3558
```
This has passed for me locally with `--repeat 10`, marked as @ignore as this depends on changes from #3558

graphcareful replied:
@ignore directive removed now that #3558 is in

graphcareful commented:
> > > I would love to add a test where nodes are being decommissioned and added with non replicable topics in the cluster. I am a bit worried about how will it operate in the face of multiple partitions movements and failures
> > 
> > I did write a test that injected failures during moves, but it failed half of the time due to the random partition assignment choosing a x-core move and thus hitting that x-core move bug we observed in dev.
> 
> @graphcareful i think i fixed the problem you describe here: #3558, do you think we can try running this test with that fix included ?
> 

It looks good to me ,I approved #3558. This PR contains the ducktape test I wrote to verify, however it's current marked as @ignore until #3558 goes in.

Awesome work!
